articles/introduction-to-avl-trees/index.md

```diff
@@ -0,0 +1,548 @@
+
+### Table Of Contents
+1. [Introduction](#introduction)
+2. [Reason for AVL Trees](#reason-for-avl-trees)
+3. [Tri-Node Restructuring](#tri-node-restructuring)
+4. [Insertion in AVL Trees in C++](#insertion-in-avl-trees-in-c)
+5. [Deletion from AVL Trees in C++](#deletion-from-avl-trees-in-c)
+6. [Conclusion](#conclusion)
+
+### Introduction
+
+AVL trees are nothing but height-balanced binary search trees. As mentioned, AVL trees follow the property of height balancing, wherein the difference of heights between the left and right nodes of a parent cannot be more than mod(1). 
+
+![AVL trees vs BST](/engineering-education/introduction-to-avl-trees/avlTree.jpg)
+
+One can observe that in figure (a), the difference between the heights of all the left and right sub-trees is less than or equal to 1. This makes the data structure represented in figure (a) an AVL tree. However, in figure (b), the height difference between the left and right children of the tree rooted at node 4 is more than 1. This makes the tree a BST and not an AVL tree. 
+
+### Reason for AVL Trees
+
+Operations on BSTs like search, insertion and deletion take O(h) time, where 'h' represents the height of the tree. In case of a skewed BST (a binary search tree where all nodes have only one or no child node), performing these operations take O(n) time, where 'n' denotes the number of nodes in the tree. To tackle this inefficiency in the process, the tree can be constructed in such a way that it always maintains a logarithmic height, thereby reducing the time complexity for all operations to O(log h).
+
+This means that the height of the tree must be maintained after every insertion and deletion such that the time complexity for every operation performed thereafter remains constant.
+
+How does one maintain the height of the tree after every such modification?
+
+### Tri-Node Restructuring
+When a node is either inserted or deleted in an AVL tree, it might create an imbalance in the tree and would result in an increase in the time complexity. But, if the tree is restructured every time there is a modification to the structure, it would ensure that the time complexity still remains O(log h).
+
+Let's see how the insertion of a node into the following tree would affect its structure.
+
+Take the example of the following AVL Tree. 
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/avl_insertion_before.jpg" alt="Before insertion"> </center>
+
+On inserting a node (4) in the tree, the structure would change to -
+
+<center> <img src="/engineering-education/introduction-to-avl-trees/tree_after_insertion.jpg" alt="After Insertion"> </center>
```

Comment:
Please include image source attributions for all the images taken from an external source.

Example:

```md
*Figure: Client Side Rendering -- [Source](https://medium.com/walmartglobaltech/the-benefits-of-server-side-rendering-over-client-side-rendering-5d07ff2cefe8)*
```

Reply:
The images weren't taken from an external source, I made all of them.

Reply:
Ah, that's great!